logstash-core/lib/logstash/config/source_loader.rb

```diff
@@ -0,0 +1,71 @@
+# encoding: utf-8
+require "logstash/config/source/local"
+require "thread"
+require "set"
+
+module LogStash module Config
+  class SourceLoader
+    class Loader
+      def initialize(loaders)
+        @loaders = loaders
+      end
+
+      def pipeline_configs
+        @loaders.collect(&:pipeline_configs).flatten
+      end
+
+      # TODO(ph) multiple pipeline
+      def pipeline_config
+        pipeline_configs.first
+      end
+    end
+
+    include LogStash::Util::Loggable
+
+    def initialize
+      @source_loaders_mutex = Mutex.new
+      @source_loaders = Set.new([LogStash::Config::Source::Local])
+    end
+
+    # This return a ConfigLoader object that will
+    # abstract the call to the different sources and will return multiples pipeline
+    def create(settings)
+      loaders = []
+
+      source_loaders do |config_source|
+        if config_source.match?(settings)
+          loaders << config_source.new(settings)
+        end
+      end
+
+      if loaders.empty?
+        # This shouldn't happen with the settings object
+        #  Or with any external plugins.
+        raise "Can't find an appropriate config loader with current settings"
```

Comment:
we should have more error information, maybe create an InvalidLoaderSetting exception or some other name?